agent/consul/leader_peering.go

```diff
@@ -294,27 +287,96 @@ func (s *Server) establishStream(ctx context.Context, logger hclog.Logger, ws me
 
 	logger.Trace("establishing stream to peer")
 
-	retryCtx, cancel := context.WithCancel(ctx)
-	cancelFns[peer.ID] = cancel
-
 	streamStatus, err := s.peerStreamTracker.Register(peer.ID)
 	if err != nil {
 		return fmt.Errorf("failed to register stream: %v", err)
 	}
 
+	streamCtx, cancel := context.WithCancel(ctx)
+	cancelFns[peer.ID] = cancel
+
+	// The following goroutine sends an up-to-date peer server address to nextServerAddr.
+	// It loads the server addresses into a ring buffer and cycles through them until:
+	//   1) streamCtx is cancelled (peer is deleted)
+	//   2) the peer is modified and the watchset fires.
+	// In case (2) we refetch the peering and rebuild the ring buffer.
+	nextServerAddr := make(chan string)
+	go func() {
+		defer close(nextServerAddr)
+
+		var (
+			init    bool
+			ringbuf *ring.Ring
+			innerWs memdb.WatchSet
+		)
+
+		fetchAddrs := func() error {
+			innerWs = memdb.NewWatchSet()
+			_, peering, err := s.fsm.State().PeeringReadByID(innerWs, peer.ID)
+			if !init {
+				init = true
+				// On the first call to `fetchAddrs` we initialize the
+				// ring buffer with the peer passed to `establishStream`
+				// (not the one fetched just above) because the caller
+				// has pre-checked `peer.ShouldDial`, guaranteeing
+				// at least one server address. Further reads from the
+				// state store may not have this guarantee.
+				ringbuf = ring.New(len(peer.PeerServerAddresses))
+				for _, addr := range peer.PeerServerAddresses {
+					ringbuf.Value = addr
+					ringbuf = ringbuf.Next()
+				}
+				return nil
```

Comment:
I might be missing something, but it looks like we can eliminate `init`,  move this out of `fetchAddrs` and replace the call to `fetchAddrs` prior to the `for` loop with this.

Reply:
We rely on the first `fetchAddrs` call to set up watches the `innerWs` as well; I could keep it separate if it makes it easier to parse. No strong opinions here

Reply:
Yeah, same. I feel like it's slightly simpler to make `fetchAddrs` have one mode, but it's essentially equivalent. I'm just leaving this comment in passing while I read peering PRs to get up to speed. Feel free to ignore this.

I think this is equivalent: 

```diff
diff --git a/agent/consul/leader_peering.go b/agent/consul/leader_peering.go
index d2b3f071e..afc6b0302 100644
--- a/agent/consul/leader_peering.go
+++ b/agent/consul/leader_peering.go
@@ -304,30 +304,15 @@ func (s *Server) establishStream(ctx context.Context, logger hclog.Logger, ws me
 	go func() {
 		defer close(nextServerAddr)
 
-		var (
-			init    bool
-			ringbuf *ring.Ring
-			innerWs memdb.WatchSet
-		)
+		innerWs := memdb.NewWatchSet()
+		ringbuf := ring.New(len(peer.PeerServerAddresses))
+		for _, addr := range peer.PeerServerAddresses {
+			ringbuf.Value = addr
+			ringbuf = ringbuf.Next()
+		}
 
 		fetchAddrs := func() error {
-			innerWs = memdb.NewWatchSet()
 			_, peering, err := s.fsm.State().PeeringReadByID(innerWs, peer.ID)
-			if !init {
-				init = true
-				// On the first call to `fetchAddrs` we initialize the
-				// ring buffer with the peer passed to `establishStream`
-				// (not the one fetched just above) because the caller
-				// has pre-checked `peer.ShouldDial`, guaranteeing
-				// at least one server address. Further reads from the
-				// state store may not have this guarantee.
-				ringbuf = ring.New(len(peer.PeerServerAddresses))
-				for _, addr := range peer.PeerServerAddresses {
-					ringbuf.Value = addr
-					ringbuf = ringbuf.Next()
-				}
-				return nil
-			}
 			if err != nil {
 				return fmt.Errorf("failed to fetch peer %q: %w", peer.ID, err)
 			}
@@ -345,11 +330,6 @@ func (s *Server) establishStream(ctx context.Context, logger hclog.Logger, ws me
 			}
 			return nil
 		}
-		if err := fetchAddrs(); err != nil {
-			s.logger.Warn("watchset for peer was fired but failed to update server addresses",
-				"peer_id", peer.ID,
-				"error", err)
-		}
 
 		for {
 			select {

```

Reply:
Your diff is missing the first call to `s.fsmState().PeeringReadByID` and its error handling (otherwise the watchset never wakes up) but I think it does read cleaner. Will push a change